src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -761,17 +862,38 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                 throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(SslStream.ReadAsync), "read"));
             }
 
+            ThrowIfExceptionalOrNotAuthenticated();
+
             Debug.Assert(_internalBuffer is null || _internalBufferCount > 0 || _decryptedBytesCount > 0, "_internalBuffer allocated when no data is buffered.");
+            int processedLength = 0;
+            int payloadBytes = 0;
 
             try
             {
-                while (true)
+
+                if (_decryptedBytesCount != 0)
                 {
-                    if (_decryptedBytesCount != 0)
+                    processedLength = CopyDecryptedData(buffer);
+                    if (_decryptedBytesCount != 0 || !haveFullTlsFrame(ref payloadBytes))
```

Comment:
It's sort of weird to test for `_decryptedBytesCount != 0` here; it wasn't obvious to me that this meant the user's buffer had been filled.

Instead I'd suggest explicitly testing for buffer.Length == 0 after doing the slice below.

Reply:
Also, it feels somewhat weird to test for `haveFullTlsFrame` here, because then we fall through to the code below which is going to potentially read more data, but it's actually not in this case because we already have enough data.

Reply:
In practice, this means that the frame was not data frame and I don't know if it is safe to keep processing. Typically this could be Alert, SessionTicket or any re/post handshake chatter. I will check if I can make it more consistent.